File: resources/src/simple_rpg_python/game_server.py
"""Python equivalent of game_server.ex.

Elixir 版の game_loop(players) を threading.Thread + queue.Queue で再現する。
Elixir の mailbox に相当する Queue を GameServer 側で1つ保有し、
プレイヤーからの (type, payload) タプルを receive 相当で取り出して分岐する。
"""

from __future__ import annotations

import queue
import threading
from typing import Optional


HAND_NAMES = {1: "Rock", 2: "Paper", 3: "Scissors"}


def _beats(hand1: int, hand2: int) -> bool:
    # Elixir の defp beats?/2 と同じ表 (1=Rock, 2=Paper, 3=Scissors)
    table = {
        (1, 2): False,
        (1, 3): True,
        (2, 3): False,
    }
    return table.get((hand1, hand2), False)


class GameServer:
    def __init__(self) -> None:
        self.mailbox: "queue.Queue[tuple]" = queue.Queue()
        self._thread: Optional[threading.Thread] = None

    def start_link(self) -> "GameServer":
        self._thread = threading.Thread(target=self._game_loop, args=({},), daemon=True)
        self._thread.start()
        return self

    def send(self, message: tuple) -> None:
        self.mailbox.put(message)

    def _game_loop(self, players: dict) -> None:
        while True:
            message = self.mailbox.get()
            match message:
                case ("join", player_mailbox, name):
                    players = {**players, name: {"mailbox": player_mailbox, "hand": None}}
                case ("play", name, hand):
                    players = {**players, name: {**players[name], "hand": hand}}
                    if self._all_hands_ready(players):
                        result = self._judge_game(players)
                        for _name, info in players.items():
                            info["mailbox"].put(("result", result))
                        players = {}
                case ("quit",):
                    return

    @staticmethod
    def _all_hands_ready(players: dict) -> bool:
        return all(info["hand"] is not None for info in players.values())

    @staticmethod
    def _judge_game(players: dict) -> str:
        hands = {name: info["hand"] for name, info in players.items()}
        hand_names = ", ".join(f"{name}: {HAND_NAMES[h]}" for name, h in hands.items())
        winners = GameServer._find_winners(hands)
        return f"Result: {hand_names} → {', '.join(winners)}"

    @staticmethod
    def _find_winners(hands: dict) -> list[str]:
        unique = sorted(set(hands.values()))
        if len(unique) == 1:
            return ["Draw!"]
        if len(unique) == 2:
            hand1, hand2 = unique
            winning_hand = hand1 if _beats(hand1, hand2) else hand2
            names = [name for name, h in hands.items() if h == winning_hand]
            return [f"{', '.join(names)} win!"]
        return ["Three-way: Scissors beats Paper, Paper beats Rock, Rock beats Scissors - Draw!"]

File: resources/src/simple_rpg/lib/game_server.ex
defmodule GameServer do
  def start_link do
    {:ok, spawn_link(fn -> game_loop(%{}) end)}
  end

  defp game_loop(players) do
    receive do
      {:join, player_pid, name} ->
        new_players = Map.put(players, name, %{pid: player_pid, hand: nil})
        game_loop(new_players)

      {:play, name, hand} ->
        updated = put_in(players[name][:hand], hand)
        check_and_announce(updated)

      :quit ->
        :ok
    end
  end

  defp check_and_announce(players) do
    if all_hands_ready?(players) do
      result = judge_game(players)
      Enum.each(players, fn {_name, %{pid: pid}} ->
        send(pid, {:result, result})
      end)
      game_loop(%{})
    else
      game_loop(players)
    end
  end

  defp all_hands_ready?(players) do
    Enum.all?(players, fn {_name, %{hand: hand}} -> hand != nil end)
  end

  defp judge_game(players) do
    hands = Map.new(players, fn {name, %{hand: hand}} -> {name, hand} end)

    hand_names = Enum.map(hands, fn {name, hand} ->
      "#{name}: #{hand_name(hand)}"
    end)
    |> Enum.join(", ")

    winners = find_winners(hands)
    winner_str = Enum.join(winners, ", ")

    "Result: #{hand_names} → #{winner_str}"
  end

  defp hand_name(hand) do
    case hand do
      1 -> "Rock"
      2 -> "Paper"
      3 -> "Scissors"
    end
  end

  defp find_winners(hands) do
    hand_values = hands |> Map.values() |> Enum.uniq()

    cond do
      length(hand_values) == 1 ->
        ["Draw!"]

      length(hand_values) == 2 ->
        [hand1, hand2] = Enum.sort(hand_values)

        if beats?(hand1, hand2) do
          winners = hands
            |> Enum.filter(fn {_name, h} -> h == hand1 end)
            |> Enum.map(fn {name, _} -> name end)
            |> Enum.join(", ")
          ["#{winners} win!"]
        else
          winners = hands
            |> Enum.filter(fn {_name, h} -> h == hand2 end)
            |> Enum.map(fn {name, _} -> name end)
            |> Enum.join(", ")
          ["#{winners} win!"]
        end

      true ->
        ["Three-way: Scissors beats Paper, Paper beats Rock, Rock beats Scissors - Draw!"]
    end
  end

  defp beats?(1, 2), do: false  # Paper beats Rock
  defp beats?(1, 3), do: true   # Rock beats Scissors
  defp beats?(2, 3), do: false  # Scissors beats Paper
  defp beats?(_, _), do: false
end
